README.md
<p align="center"><img src="assets/horizontal.png" alt="neuron" height="150px"></p>

[![Build Status](https://github.com/uesteibar/neuron/actions/workflows/ci.yml/badge.svg)](https://github.com/uesteibar/neuron/actions/workflows/ci.yml)
[![Hex Version](https://img.shields.io/hexpm/v/neuron.svg)](https://hex.pm/packages/neuron)

A GraphQL client for Elixir.

## Index

- [Installation](#installation)
- [Usage](#usage)
- [Running Locally](#running-locally)
- [Contributing](#contributing)

## Installation

```elixir
def deps do
  [
    {:neuron, "~> 5.1.0"}
  ]
end
```

## JSON library

Neuron defaults to using Jason for JSON encoding and decoding. To use Jason, add it to your deps:

```elixir
{:jason, "~> 1.1"}
```

It is also possible to customize which JSON library that is used:

```elixir
Neuron.Config.set(json_library: AnotherJSONLibrary)
```

## Connection

Neuron defaults to using HTTP(S) protocol with HTTPoison for Connecting to GraphQL endpoint. You can however customize that behaviour, by providing custom library, which should implement Neuron.Connection behaviour:

```elixir
defmodule MyConnection do
  @behaviour Neuron.Connection

  @impl Neuron.Connection
  def call(body, options) do
    IO.inspect("NEURON CALLED")
    Neuron.Connection.Http.call(body, options)
  end
end
```

Then set it up in config:

```elixir
Neuron.Config.set(connection_module: MyConnection)
```

## Usage

```elixir
iex> Neuron.Config.set(url: "https://example.com/graph")

iex> Neuron.query("""
      {
        films {
          count
        }
      }
    """)
```

Response will be:

```elixir
{:ok, %Neuron.Response{body: %{"data" => %{"films" => %{ "count": 123 }}}, status_code: 200, headers: []}}
```

You can also run mutations:

```elixir
iex> Neuron.query("""
      mutation createUser($name: String!) {
        createUser(name: $name) {
          id
          name
        }
      }
    """,
    %{name: "uesteibar"}
    )
```

You can also set url and headers as shown below:

```elixir
iex> Neuron.query("""
      mutation createUser($name: String!) {
        createUser(name: $name) {
          id
          name
        }
      }
    """,
    %{name: "uesteibar"},
    url: "https://example.com/graph",
    headers: [authorization: "Bearer <token>"]
    )
```

### Overriding HTTP Timeout
`HTTPoison` default timeout is 5000ms, in case we need to handle longer timeout, using default `Neuron.Connection` module, we could set `connection_opts` which will be passed to `HTTPoison`. So to override timeout to 15000ms, we could do:

```elixir
iex> Neuron.Config.set(url: "https://example.com/graph", connection_opts: [recv_timeout: 15_000])

iex> Neuron.query("""
      {
        films {
          count
        }
      }
    """)
```

We can also set the timeout for a single request by passing the `connection_opts` to `Neuron.query/3` instead:

```elixir
iex> Neuron.query("...", %{}, connection_opts: [recv_timeout: 15_000])
```

More extensive documentation can be found at [https://hexdocs.pm/neuron](https://hexdocs.pm/neuron).

## Running locally

Clone the repository:

```bash
git clone git@github.com:uesteibar/neuron.git
```

Install dependencies:

```bash
cd neuron
mix deps.get
```

To run the tests:

```bash
mix test
```

## Style guide

Code is formatted with `mix format` and `mix credo` should not show warnings.

To format the code and run static code analysis with credo

```elixir
mix format
mix credo
```

## Contributing

Pull requests are always welcome =)

The project uses [standard-changelog](https://github.com/conventional-changelog/conventional-changelog) to update the [Changelog](https://github.com/uesteibar/neuron/blob/master/CHANGELOG.md) with each commit message and upgrade the package version.
For that reason every contribution should have a title and body that follows the [conventional commits standard](https://conventionalcommits.org/) conventions (e.g. `feat(connection): Make it smarter than Jarvis`).

To make this process easier, you can do the following:

Install `commitizen` and `cz-conventional-changelog` globally:

```bash
npm i -g commitizen cz-conventional-changelog
```

Save `cz-conventional-changelog` as default:

```bash
echo '{ "path": "cz-conventional-changelog" }' > ~/.czrc
```

Instead of `git commit`, you can now run:

```
git cz
```

and follow the instructions to generate the commit message.

## Copyright and License

Copyright (c) 2017 Unai Esteibar

This software is released under the [Internet Systems Consortium License](./LICENSE.md).


CHANGELOG.md
# Changelog

<a name="5.1.0"></a>
## [5.1.0](https://github.com/uesteibar/neuron/compare/v5.0.0...v5.1.0) (2023-02-22)

* Support HTTPoison 2.0 (#74) ([da19f70](https://github.com/uesteibar/neuron/commit/da19f70)), closes [#74](https://github.com/uesteibar/neuron/issues/74)

### Bug Fixes

* ci(github): bump deps and switch to github actions (#72) ([c56bc74](https://github.com/uesteibar/neuron/commit/c56bc74)), closes [#72](https://github.com/uesteibar/neuron/issues/72)



<a name="5.0.0"></a>
## [5.0.0](https://github.com/uesteibar/neuron/compare/v4.1.2...v5.0.0) (2020-06-10)


### Bug Fixes

* **neuron.connection.http:** fix error case so return value matches typespec ([#60](https://github.com/uesteibar/neuron/issues/60)) ([b96d871](https://github.com/uesteibar/neuron/commit/b96d87189bb2e8194e905d0cbc7e62199d2594c3)), closes [#59](https://github.com/uesteibar/neuron/issues/59)


### BREAKING CHANGES

* **neuron.connection.http:** If your code expects the `response` field of `JSONParseError` to be a tuple, it will break.



<a name="4.1.2"></a>
## [4.1.2](https://github.com/uesteibar/neuron/compare/v4.1.1...v4.1.2) (2020-02-19)


### Bug Fixes

* **warning:**  compile time warning (default arguments in handle/3 are never used) ([75c151c](https://github.com/uesteibar/neuron/commit/75c151c))



<a name="4.1.1"></a>
## [4.1.1](https://github.com/uesteibar/neuron/compare/v4.1.0...v4.1.1) (2019-12-26)


### Bug Fixes

* **fragments:** remove duplicate fragments when inserting them into the query ([d3a5c3d](https://github.com/uesteibar/neuron/commit/d3a5c3d)), closes [#56](https://github.com/uesteibar/neuron/issues/56)



<a name="4.1.0"></a>
## [4.1.0](https://github.com/uesteibar/neuron/compare/v4.0.0...v4.1.0) (2019-11-21)


### Features

* **connection:** Allow to inject custom connection module ([f22287e](https://github.com/uesteibar/neuron/commit/f22287e))



<a name="4.0.0"></a>
## [4.0.0](https://github.com/uesteibar/neuron/compare/v3.0.1...v4.0.0) (2019-10-25)


### Bug Fixes

* **connection:** fix function signature for Neuron.Connection.post ([82817e7](https://github.com/uesteibar/neuron/commit/82817e7))
* **typespecs:** Fix typespecs for Neuron.query/3 and Neuron.Store.set/3 ([263fef4](https://github.com/uesteibar/neuron/commit/263fef4))


### Features

* **fragment:** raise `Neuron.MissingFragmentsError` for queries with missing fragments ([75ddf54](https://github.com/uesteibar/neuron/commit/75ddf54))


### BREAKING CHANGES

* **fragment:** Queries with missing fragments will now raise `Neuron.MissingFragmentsError` instead of a regular exception.



# Unreleased

* **typespecs:** Fix various type specs, specifically Neuron.query/2 can return a HTTPoison.Error.t().

<a name="3.0.1"></a>
## [3.0.1](https://github.com/uesteibar/neuron/compare/v3.0.0...v3.0.1) (2019-08-26)


### Bug Fixes

* **deps:** Make poison a test dependency ([d75e10b](https://github.com/uesteibar/neuron/commit/d75e10b))



<a name="3.0.0"></a>
## [3.0.0](https://github.com/uesteibar/neuron/compare/v2.0.0...v3.0.0) (2019-08-23)


### Features

* **json:** Use Jason as default JSOn library ([3b4e472](https://github.com/uesteibar/neuron/commit/3b4e472))


### BREAKING CHANGES

* **json:** JSON parsing library is now [Jason](https://github.com/michalmuskala/jason).
In order to keep using [Poison](https://github.com/devinus/poison)
you'll need to manually add it as a dependency and configure neuron to
use it by running `Neuron.Config.set(json_library: AnotherJSONLibrary)`.



<a name="2.0.0"></a>
## [2.0.0](https://github.com/uesteibar/neuron/compare/v1.2.0...v2.0.0) (2019-06-26)


### chore

* **deps:** Update poison to 4.0, mock to 0.3.3 and coverex to 1.5 ([600db92](https://github.com/uesteibar/neuron/commit/600db92))


### BREAKING CHANGES

* Now requires elixir >= 1.6



<a name="1.2.0"></a>
## [1.2.0](https://github.com/uesteibar/neuron/compare/v1.1.1...v1.2.0) (2019-05-30)


### Features

* **fragment:** recursively embed fragments ([3807d89](https://github.com/uesteibar/neuron/commit/3807d89))



<a name="1.1.1"></a>
## [1.1.1](https://github.com/uesteibar/neuron/compare/v1.1.0...v1.1.1) (2019-03-14)


### Bug Fixes

* **typespecs:** Fix typespec for `Neuron.query/3` ([e267f8c](https://github.com/uesteibar/neuron/commit/e267f8c)), closes [#31](https://github.com/uesteibar/neuron/issues/31)



<a name="1.1.0"></a>
## [1.1.0](https://github.com/uesteibar/neuron/compare/v1.0.0...v1.1.0) (2018-12-25)


### Features

* **decode:** Add support for passing options to the json decoder ([#29](https://github.com/uesteibar/neuron/issues/29)) ([f8425ac](https://github.com/uesteibar/neuron/commit/f8425ac))
* **options:** Allow to override connection options per request ([#30](https://github.com/uesteibar/neuron/issues/30)) ([30d99e5](https://github.com/uesteibar/neuron/commit/30d99e5))
* **response:** return a meaningful error when response is not JSON ([#27](https://github.com/uesteibar/neuron/issues/27)) ([32004b2](https://github.com/uesteibar/neuron/commit/32004b2))



<a name="1.0.0"></a>
## [1.0.0](https://github.com/uesteibar/neuron/compare/v0.9.1...v1.0.0) (2018-10-15)


### Features

* **query:** Allow passing variables ([ff82f0a](https://github.com/uesteibar/neuron/commit/ff82f0a)), closes [#25](https://github.com/uesteibar/neuron/issues/25)


### BREAKING CHANGES

  - `mutation/1` and `mutation/2` are deprecated
  - `query/2` is replaced by `query/3` to allow passing variables in



<a name="0.9.1"></a>
## [0.9.1](https://github.com/uesteibar/neuron/compare/v0.9.0...v0.9.1) (2018-09-21)


### Bug Fixes

* **headers:** Use `application/json` content type with :as_json ([f369555](https://github.com/uesteibar/neuron/commit/f369555))



<a name="0.9.0"></a>
# [0.9.0](https://github.com/uesteibar/neuron/compare/v0.8.0...v0.9.0) (2018-09-08)


### Features


mix.exs
defmodule Neuron.Mixfile do
  use Mix.Project

  @version "5.1.0"
  @github "https://github.com/uesteibar/neuron"

  def project do
    [
      app: :neuron,
      description: "A GraphQL client for elixir.",
      version: @version,
      elixir: "~> 1.7",
      build_embedded: Mix.env() == :prod,
      start_permanent: Mix.env() == :prod,
      test_coverage: [tool: Coverex.Task],
      aliases: aliases(),
      package: package(),
      deps: deps(),
      docs: docs(),
      dialyzer: [
        plt_add_deps: :transitive,
        plt_add_apps: [:mix],
        plt_file: {:no_warn, "priv/plts/dialyzer.plt"}
      ]
    ]
  end

  def application do
    [extra_applications: [:logger, :httpoison]]
  end

  defp deps do
    [
      {:httpoison, "~> 1.0 or ~> 2.0"},
      {:coverex, "~> 1.5", only: :test},
      {:credo, "~> 1.1", only: [:dev, :test]},
      {:dialyxir, "~> 1.0", only: [:dev, :test], runtime: false},
      {:ex_doc, ">= 0.0.0", only: :dev, runtime: false},
      {:jason, "~> 1.1", optional: true},
      {:mock, "~> 0.3.3", only: :test},
      {:poison, "~> 4.0", only: :test}
    ]
  end

  defp package do
    [
      files: [
        "lib",
        "mix.exs",
        "README.md",
        "LICENSE.md"
      ],
      links: %{"github" => @github},
      maintainers: ["Unai Esteibar <uesteibar@gmail.com>"],
      licenses: ["ISC"]
    ]
  end

  defp docs do
    [
      extras: [
        "CHANGELOG.md": [],
        "LICENSE.md": [title: "License"],
        "README.md": [title: "Overview"]
      ],
      main: "readme",
      assets: "assets",
      source_url: @github,
      source_ref: "v#{@version}",
      logo: "assets/logo.png",
      formatters: ["html"]
    ]
  end

  defp aliases do
    [
      test: "test --cover"
    ]
  end
end
